File: tools/check_adrs.py
#!/usr/bin/env python3
"""Docs lint (T3): ADR numbering/status sanity + relative-link resolution.

Run: python3 tools/check_adrs.py   (exit 0 = ok, 1 = problems found)
"""
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DOCS = ROOT / "docs"
ADR = DOCS / "adr"
LINK_RE = re.compile(r"\[[^\]]+\]\(([^)]+)\)")
STATUS_RE = re.compile(r"状态[:：]")


def check_adrs(errors):
    seen = {}
    for fp in sorted(ADR.glob("*.md")):
        m = re.match(r"(\d{4})-.*\.md$", fp.name)
        if not m:
            errors.append(f"ADR filename not NNNN-*.md: {fp.name}")
            continue
        num = m.group(1)
        if num in seen:
            errors.append(f"duplicate ADR number {num}: {fp.name} & {seen[num]}")
        seen[num] = fp.name
        text = fp.read_text(encoding="utf-8")
        if not STATUS_RE.search(text):
            errors.append(f"{fp.name}: missing 状态 field")
    if not seen:
        errors.append("no ADR files found")


def check_links(errors):
    for fp in sorted(DOCS.rglob("*.md")):
        for target in LINK_RE.findall(fp.read_text(encoding="utf-8")):
            t = target.strip()
            if t.startswith(("http://", "https://", "#", "mailto:")):
                continue
            t = t.split("#", 1)[0].split("?", 1)[0]
            if not t:
                continue
            if not (fp.parent / t).resolve().exists():
                errors.append(f"{fp.relative_to(ROOT)}: broken link -> {target}")


def main():
    errors = []
    check_adrs(errors)
    check_links(errors)
    if errors:
        print("docs lint FAILED:")
        for e in errors:
            print("  -", e)
        return 1
    print("docs lint OK: ADR numbering/status valid, all relative links resolve.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/check_adrs.mjs
#!/usr/bin/env node
// Docs lint: ADR numbering/status sanity + relative-link resolution.
// Run: node tools/check_adrs.mjs   (exit 0 = ok, 1 = problems found)
//
// Pure Node ≥18 (ESM, no external deps).
import { readFileSync, readdirSync, existsSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import path from 'node:path'

const HERE = path.dirname(fileURLToPath(import.meta.url))
const ROOT = path.resolve(HERE, '..')
const DOCS = path.join(ROOT, 'docs')
const ADR = path.join(DOCS, 'adr')
const LINK_RE = /\[[^\]]+\]\(([^)]+)\)/g
const STATUS_RE = /状态[:：]/

function walkMd(dir) {
  const out = []
  for (const e of readdirSync(dir, { withFileTypes: true })) {
    const fp = path.join(dir, e.name)
    if (e.isDirectory()) out.push(...walkMd(fp))
    else if (e.name.endsWith('.md')) out.push(fp)
  }
  return out
}

function checkAdrs(errors) {
  const seen = new Map()
  const files = readdirSync(ADR)
    .filter((n) => n.endsWith('.md'))
    .sort()
  for (const name of files) {
    const m = name.match(/^(\d{4})-.*\.md$/)
    if (!m) {
      errors.push(`ADR filename not NNNN-*.md: ${name}`)
      continue
    }
    const num = m[1]
    if (seen.has(num)) errors.push(`duplicate ADR number ${num}: ${name} & ${seen.get(num)}`)
    seen.set(num, name)
    const text = readFileSync(path.join(ADR, name), 'utf8')
    if (!STATUS_RE.test(text)) errors.push(`${name}: missing 状态 field`)
  }
  if (seen.size === 0) errors.push('no ADR files found')
}

/** Strip fenced code blocks so example links inside ``` aren't checked. */
function stripFencedBlocks(text) {
  return text.replace(/^```[^\n]*\n[\s\S]*?^```/gm, '')
}

function checkLinks(errors) {
  for (const fp of walkMd(DOCS).sort()) {
    const text = stripFencedBlocks(readFileSync(fp, 'utf8'))
    for (const match of text.matchAll(LINK_RE)) {
      const target = match[1].trim()
      if (/^(https?:\/\/|#|mailto:)/.test(target)) continue
      const t = target.split('#')[0].split('?')[0]
      if (!t) continue
      if (!existsSync(path.resolve(path.dirname(fp), t))) {
        errors.push(`${path.relative(ROOT, fp)}: broken link -> ${target}`)
      }
    }
  }
}

const errors = []
checkAdrs(errors)
checkLinks(errors)
if (errors.length) {
  console.log('docs lint FAILED:')
  for (const e of errors) console.log('  -', e)
  process.exit(1)
}
console.log('docs lint OK: ADR numbering/status valid, all relative links resolve.')
